interoperator/kubeconfig_rotation/main.go

```diff
@@ -0,0 +1,209 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	authenticationv1alpha1 "github.com/gardener/gardener/pkg/apis/authentication/v1alpha1"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	cl "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var sa_sec *corev1.Secret = &corev1.Secret{}
+var secList *corev1.SecretList = &corev1.SecretList{}
+var shoot *gardencorev1beta1.Shoot = &gardencorev1beta1.Shoot{}
+var ms_cl cl.Client
+var sa_cl cl.Client
+var ctx context.Context
+var adminKubeconfigRequest *authenticationv1alpha1.AdminKubeconfigRequest = &authenticationv1alpha1.AdminKubeconfigRequest{}
+var updated_sec *corev1.Secret = &corev1.Secret{}
+var in_cluster_config *rest.Config
+
+func getInClusterConfig() (*rest.Config, error) {
+	in_cluster_config, err := rest.InClusterConfig()
+	return in_cluster_config, err
+}
+
+type RetryableOperation func() error
+
+func getServiceAccountKubeconfigSecret() error {
+	err := ms_cl.Get(context.Background(), cl.ObjectKey{
+		Name:      "service-account-kubeconfig",
+		Namespace: os.Getenv("NAMESPACE"),
+	}, sa_sec)
+	return err
+}
+
+func getSecretList() error {
+	err := ms_cl.List(ctx, secList)
+	return err
+}
+
+func getShootCluster(shoot_namespace string, shoot_name string) error {
+	err := sa_cl.Get(ctx, types.NamespacedName{Namespace: shoot_namespace, Name: shoot_name}, shoot)
+	return err
+}
+
+func createAdminKubeConfig() error {
+	err := sa_cl.SubResource("adminkubeconfig").Create(ctx, shoot, adminKubeconfigRequest)
+	return err
+}
+
+func updateShootSecret(secret_name string, secret_namespace string) error {
+	err1 := ms_cl.Get(context.Background(), cl.ObjectKey{
+		Name:      secret_name,
+		Namespace: secret_namespace,
+	}, updated_sec)
+	if err1 != nil {
+		return err1
+	}
+	shoot_kubeconfig := adminKubeconfigRequest.Status.Kubeconfig
+	updated_sec.Data["kubeconfig"] = shoot_kubeconfig
+	err2 := ms_cl.Update(ctx, updated_sec)
+	return err2
+}
+
+func Retry(maxRetries int, sleep time.Duration, operation RetryableOperation) error {
+	var err error
+	for attempt := 1; attempt <= maxRetries; attempt++ {
+		err = operation()
+		if err == nil {
+			return nil
+		}
+		fmt.Println(err)
+		fmt.Println("Retrying...")
+		time.Sleep(sleep * time.Duration(attempt))
+	}
+	fmt.Println(err)
+	fmt.Println("All retries failed.")
```

Comment:
Please use https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/log for logging